File: Chapter_3/web.js
/**
 * web.js - Clean web interface for JavaScript LLM testers
 * Provides clean JSON responses without CLI artifacts
 */

import express from 'express';
import cors from 'cors';
import { getDisplayName, getDefaultModel } from './utils.js';

// Capture console output for web responses
function captureConsoleOutput(fn) {
    const originalLog = console.log;
    const logs = [];
    
    console.log = (...args) => {
        logs.push(args.join(' '));
    };
    
    try {
        const result = fn();
        return { result, logs: logs.join('\n') };
    } finally {
        console.log = originalLog;
    }
}

async function captureConsoleOutputAsync(fn) {
    const originalLog = console.log;
    const logs = [];
    
    console.log = (...args) => {
        logs.push(args.join(' '));
    };
    
    try {
        const result = await fn();
        return { result, logs: logs.join('\n') };
    } finally {
        console.log = originalLog;
    }
}

function createWebApi(ManagerClass) {
    const app = express();
    
    // Middleware
    app.use(cors());
    app.use(express.json());
    
    // Initialize the manager
    let manager;
    
    // Initialize manager asynchronously
    const initPromise = (async () => {
        manager = new ManagerClass();
        await manager._checkProviders();
        return manager;
    })();
    
    // Middleware to ensure manager is initialized
    app.use(async (req, res, next) => {
        if (!manager) {
            await initPromise;
        }
        next();
    });
    
    // Get service status
    app.get('/', async (req, res) => {
        try {
            const availableProviders = manager.getAvailableProviders();
            
            res.json({
                framework: manager.framework,
                available_providers: availableProviders,
                total_available: availableProviders.length,
                initialization_status: manager.initializationMessages,
                status: availableProviders.length > 0 ? 'healthy' : 'no_providers'
            });
        } catch (error) {
            res.status(500).json({ error: error.message });
        }
    });
    
    // Get available providers with details
    app.get('/providers', async (req, res) => {
        try {
            const availableProviders = manager.getAvailableProviders();
            
            const providersDetail = availableProviders.map(provider => ({
                name: provider,
                display_name: getDisplayName(provider),
                model: getDefaultModel(provider),
                status: manager.initializationMessages[provider] || 'Unknown'
            }));
            
            res.json({
                framework: manager.framework,
                providers: providersDetail,
                count: availableProviders.length
            });
        } catch (error) {
            res.status(500).json({ error: error.message });
        }
    });
    
    // Query single provider
    app.post('/query', async (req, res) => {
        try {
            const {
                topic,
                provider = null,
                template = '{topic}',
                max_tokens = 1000,
                temperature = 0.7
            } = req.body;
            
            if (!topic) {
                return res.status(400).json({ error: 'Topic is required' });
            }
            
            // Capture any console output
            const { result, logs } = await captureConsoleOutputAsync(async () => {
                return await manager.askQuestion(topic, provider, template, max_tokens, temperature);
            });
            
            if (!result.success) {
                return res.status(400).json({
                    error: result.error || 'Query failed',
                    provider: result.provider,
                    debug: logs || null
                });
            }
            
            // Clean response for web
            const cleanResult = {
                success: true,
                framework: manager.framework,
                provider: result.provider,
                model: result.model,
                response: result.response,
                parameters: {
                    temperature: result.temperature,
                    max_tokens: result.maxTokens,
                    template: template
                },
                prompt: result.prompt
            };
            
            // Optionally include debug info
            if (logs) {
                cleanResult.debug = logs;
            }
            
            res.json(cleanResult);
            
        } catch (error) {
            res.status(500).json({
                error: error.message,
                framework: manager.framework
            });
        }
    });
    
    // Query all available providers
    app.post('/query-all', async (req, res) => {
        try {
            const {
                topic,
                template = '{topic}',
                max_tokens = 1000,
                temperature = 0.7
            } = req.body;
            
            if (!topic) {
                return res.status(400).json({ error: 'Topic is required' });
            }
            
            // Capture any console output
            const { result, logs } = await captureConsoleOutputAsync(async () => {
                return await manager.queryAllProviders(topic, template, max_tokens, temperature);
            });
            
            if (!result.success) {
                return res.status(400).json({
                    error: result.error || 'Query failed',
                    framework: manager.framework,
                    debug: logs || null
                });
            }
            
            // Clean up the responses for web
            const cleanResponses = {};
            let successful = 0;
            let failed = 0;
            
            for (const [provider, response] of Object.entries(result.responses)) {
                if (response.success) {
                    cleanResponses[provider] = {
                        success: true,
                        response: response.response,
                        model: response.model,
                        parameters: {
                            temperature: response.temperature,
                            max_tokens: response.maxTokens
                        }
                    };
                    successful++;
                } else {
                    cleanResponses[provider] = {
                        success: false,
                        error: response.error || 'Unknown error',
                        model: response.model || 'unknown'
                    };
                    failed++;
                }
            }
            
            const cleanResult = {
                success: true,
                framework: manager.framework,
                prompt: result.prompt,
                responses: cleanResponses,
                summary: {
                    total_providers: Object.keys(result.responses).length,
                    successful: successful,
                    failed: failed
                },
                parameters: {
                    temperature: temperature,
                    max_tokens: max_tokens,
                    template: template
                }
            };
            
            // Optionally include debug info
            if (logs) {
                cleanResult.debug = logs;
            }
            
            res.json(cleanResult);
            
        } catch (error) {
            res.status(500).json({
                error: error.message,
                framework: manager.framework
            });
        }
    });
    
    // Simple health check
    app.get('/health', async (req, res) => {
        try {
            const availableProviders = manager.getAvailableProviders();
            res.json({
                status: availableProviders.length > 0 ? 'healthy' : 'unhealthy',
                framework: manager.framework,
                providers_available: availableProviders.length
            });
        } catch (error) {
            res.status(500).json({ error: error.message });
        }
    });
    
    return app;
}

export async function runWebServer(ManagerClass, host = '0.0.0.0', port = 8000) {
    const app = createWebApi(ManagerClass);
    
    // Get framework name for display
    let frameworkName = 'Unknown';
    try {
        const tempManager = new ManagerClass();
        frameworkName = tempManager.framework;
    } catch (error) {
        // Ignore
    }
    
    app.listen(port, host, () => {
        console.log(`Starting web server for ${frameworkName} framework...`);
        console.log(`API documentation available at: http://${host}:${port}/`);
        console.log(`Health check: http://${host}:${port}/health`);
        console.log(`Status: http://${host}:${port}/`);
        console.log('Press Ctrl+C to stop the server');
    });
}

// Main function for standalone web server
function main() {
    console.log('Universal LLM Web API (JavaScript)');
    console.log('This should be called from an LLM tester script like:');
    console.log('node llm_tester.js web');
}

// Run if this is the main module
if (import.meta.url === `file://${process.argv[1]}`) {
    main();
}

File: Chapter_3/web.py
"""
web.py - Clean web interface for LLM testers
Provides clean JSON responses without CLI artifacts
"""

from fastapi import FastAPI, HTTPException
from fastapi.middleware.cors import CORSMiddleware
from pydantic import BaseModel
from typing import Optional, Dict, Any
import uvicorn
import sys
import io
from contextlib import redirect_stdout

# Pydantic models for API
class QueryRequest(BaseModel):
    topic: str
    provider: Optional[str] = None
    template: str = "{topic}"
    max_tokens: int = 1000
    temperature: float = 0.7

class QueryAllRequest(BaseModel):
    topic: str
    template: str = "{topic}"
    max_tokens: int = 1000
    temperature: float = 0.7

def create_web_api(manager_class):
    """Create a web API for any LLM manager class"""
    
    app = FastAPI(
        title="LLM Service API", 
        version="1.0.0",
        description="Universal API for LLM framework testing"
    )
    
    # CORS middleware
    app.add_middleware(
        CORSMiddleware,
        allow_origins=["*"],
        allow_credentials=True,
        allow_methods=["*"],
        allow_headers=["*"],
    )
    
    # Initialize the manager
    manager = manager_class()
    
    @app.get("/")
    async def get_status():
        """Get service status"""
        available_providers = manager.get_available_providers()
        
        return {
            "framework": manager.framework,
            "available_providers": available_providers,
            "total_available": len(available_providers),
            "initialization_status": manager.initialization_messages,
            "status": "healthy" if available_providers else "no_providers"
        }
    
    @app.get("/providers")
    async def get_providers():
        """Get available providers with details"""
        from utils import get_display_name, get_default_model
        
        available_providers = manager.get_available_providers()
        
        providers_detail = []
        for provider in available_providers:
            providers_detail.append({
                "name": provider,
                "display_name": get_display_name(provider),
                "model": get_default_model(provider),
                "status": manager.initialization_messages.get(provider, "Unknown")
            })
        
        return {
            "framework": manager.framework,
            "providers": providers_detail,
            "count": len(available_providers)
        }
    
    @app.post("/query")
    async def query_single(request: QueryRequest):
        """Query a single provider - returns clean JSON"""
        try:
            # Capture any print statements from the manager
            captured_output = io.StringIO()
            
            with redirect_stdout(captured_output):
                result = manager.ask_question(
                    topic=request.topic,
                    provider=request.provider,
                    template=request.template,
                    max_tokens=request.max_tokens,
                    temperature=request.temperature
                )
            
            # Get any debug output for optional inclusion
            debug_output = captured_output.getvalue().strip()
            
            if not result.get("success", False):
                raise HTTPException(
                    status_code=400, 
                    detail={
                        "error": result.get("error", "Query failed"),
                        "provider": result.get("provider"),
                        "debug": debug_output if debug_output else None
                    }
                )
            
            # Clean response for web
            clean_result = {
                "success": True,
                "framework": manager.framework,
                "provider": result["provider"],
                "model": result["model"],
                "response": result["response"],
                "parameters": {
                    "temperature": result["temperature"],
                    "max_tokens": result["max_tokens"],
                    "template": request.template
                },
                "prompt": result["prompt"]
            }
            
            # Optionally include debug info
            if debug_output:
                clean_result["debug"] = debug_output
            
            return clean_result
            
        except HTTPException:
            raise
        except Exception as e:
            raise HTTPException(
                status_code=500, 
                detail={
                    "error": str(e),
                    "framework": manager.framework
                }
            )
    
    @app.post("/query-all")
    async def query_all(request: QueryAllRequest):
        """Query all available providers - returns clean JSON"""
        try:
            # Capture any print statements
            captured_output = io.StringIO()
            
            with redirect_stdout(captured_output):
                result = manager.query_all_providers(
                    topic=request.topic,
                    template=request.template,
                    max_tokens=request.max_tokens,
                    temperature=request.temperature
                )
            
            debug_output = captured_output.getvalue().strip()
            
            if not result.get("success", False):
                raise HTTPException(
                    status_code=400, 
                    detail={
                        "error": result.get("error", "Query failed"),
                        "framework": manager.framework,
                        "debug": debug_output if debug_output else None
                    }
                )
            
            # Clean up the responses for web
            clean_responses = {}
            for provider, response in result["responses"].items():
                if response.get("success"):
                    clean_responses[provider] = {
                        "success": True,
                        "response": response["response"],
                        "model": response["model"],
                        "parameters": {
                            "temperature": response["temperature"],
                            "max_tokens": response["max_tokens"]
                        }
                    }
                else:
                    clean_responses[provider] = {
                        "success": False,
                        "error": response.get("error", "Unknown error"),
                        "model": response.get("model", "unknown")
                    }
            
            clean_result = {
                "success": True,
                "framework": manager.framework,
                "prompt": result["prompt"],
                "responses": clean_responses,
                "summary": {
                    "total_providers": len(result["responses"]),
                    "successful": len([r for r in clean_responses.values() if r["success"]]),
                    "failed": len([r for r in clean_responses.values() if not r["success"]])
                },
                "parameters": {
                    "temperature": request.temperature,
                    "max_tokens": request.max_tokens,
                    "template": request.template
                }
            }
            
            # Optionally include debug info
            if debug_output:
                clean_result["debug"] = debug_output
            
            return clean_result
            
        except HTTPException:
            raise
        except Exception as e:
            raise HTTPException(
                status_code=500, 
                detail={
                    "error": str(e),
                    "framework": manager.framework
                }
            )
    
    @app.get("/health")
    async def health_check():
        """Simple health check"""
        available_providers = manager.get_available_providers()
        return {
            "status": "healthy" if available_providers else "unhealthy",
            "framework": manager.framework,
            "providers_available": len(available_providers)
        }
    
    return app

def run_web_server(manager_class, host: str = "0.0.0.0", port: int = 8000):
    """Run the web server with the given manager class"""
    app = create_web_api(manager_class)
    
    # Get framework name for display
    framework_name = "Unknown"
    try:
        framework_name = manager_class().framework
    except:
        pass
    
    print(f"Starting web server for {framework_name} framework...")
    print(f"API documentation available at: http://{host}:{port}/docs")
    print(f"Health check: http://{host}:{port}/health")
    print(f"Status: http://{host}:{port}/")
    
    uvicorn.run(app, host=host, port=port)

def main():
    """
    Main function for standalone web server
    Usage: python web.py
    """
    print("Universal LLM Web API")
    print("This should be called from an LLM tester script like:")
    print("python llm_tester.py web")

if __name__ == "__main__":
    main()
